src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,333 @@
+import field_theory.finite.trace
+
+/-!
+# Additive characters of finite rings and fields
+
+Let `R` be a finite commutative ring. An *additive character* of `R` with values
+in another commutative ring `R'` is simply a morphism from the additive group
+of `R` into the multiplicative monoid of `R'`.
+
+We use the namespace `add_char`.
+
+## Main definitions and results
+
+We define `mul_shift ψ a`, where `ψ : add_char R R'` and `a : R`, to be the
+character defined by `x ↦ ψ (a * x)`. An additive character `ψ` is *primitive*
+if `mul_shift ψ a` is trivial only when `a = 0`.
+
+We show that when `ψ` is primitive, then the map `a ↦ mul_shift ψ a` is injective
+(`to_mul_shift_inj_of_is_primitive`) and that `ψ` is primitive when `R` is a field
+and `ψ` is nontrivial.
+
+We also show that there are primitive additive characters on `R` (with suitable
+target `R'`) when `R` is a field or `R = zmod n` (`primitive_char_finite_field`
+and `primitive_zmod_char`).
+
+Finally, we show that the sum of all character values is zero when the character
+is nontrivial (and the target is a domain); see `sum_eq_zero_of_is_nontrivial`.
+
+## Tags
+
+additive character
+-/
+
+/-!
+### Definitions related to and results on additive characters
+-/
+
+section add_char_def
+
+universes u v
+
+-- The domain of our additive characters
+variables (R : Type u) [comm_ring R]
+-- The target
+variables (R' : Type v) [comm_ring R']
+
+/-- Define `add_char R R'` as `(multiplicative R) →* R'`. -/
+abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+
+end add_char_def
+
+
+section additive
+
+namespace add_char
+
+open multiplicative
+
+universes u v
+
+-- The domain of our additive characters
+variables {R : Type u} [comm_ring R]
+-- The target
+variables {R' : Type v} [comm_ring R']
+
+/-- The trivial additive character; it sends everything to `1`. -/
+@[simps] protected
+def trivial : add_char R R' :=
+{ to_fun := λ x, 1,
+  map_one' := rfl,
+  map_mul' := λ x y, (mul_one 1).symm }
+
+/-- An additive character is *nontrivial* if it takes a value `≠ 1`. -/
+def is_nontrivial (ψ : add_char R R') : Prop := ∃ (a : R), ψ a ≠ 1
+
+/-- An additive character is nontrivial iff it is not the trivial character. -/
+lemma is_nontrivial_iff_ne_trivial (ψ : add_char R R') :
+  is_nontrivial ψ ↔ ψ ≠ add_char.trivial :=
+begin
+  refine ⟨λ ⟨a, ha⟩ hf, ne_of_eq_of_ne (trivial_apply a).symm
+                         (ne_of_eq_of_ne (congr_fun (congr_arg coe_fn hf) a).symm ha) rfl, _⟩,
+  contrapose!,
+  intro h,
+  simp only [is_nontrivial, not_exists_not] at h,
+  ext,
+  exact (h x).trans (trivial_apply x).symm,
+end
+
+/-- Define the multiplicative shift of an additive character.
+This satisfies `mul_shift ψ a x = ψ (a * x)`. -/
+def mul_shift (ψ : add_char R R') (a : R) : add_char R R' :=
+ψ.comp (add_monoid_hom.mul_left a).to_multiplicative
+
+lemma mul_shift_spec (ψ : add_char R R') (a x : R) : mul_shift ψ a x = ψ (of_add (a * x)) := rfl
```

Comment:
```suggestion
lemma mul_shift_spec {ψ : add_char R R'} {a : R} {x : multiplicative R} : mul_shift ψ a x = ψ (of_add (a * x)) := rfl
```

Reply:
What is the advantage of this?

Reply:
If you're using `@[simps]` on `mul_shift` then you don't need `mul_shift_spec`, right? Note that my suggested fix is also wrong. You are actively abusing definitional equality in theorem statements. For me this is a bad idea. The problem with continually telling the system that you have `a : R` which is also `a : multiplicative R` is that eventually you end up with terms that display like `a * b` but which are actually `a + b` and this can be very confusing for users -- people talk of type equality "leaking", and this leakage is caused by statements which abuse defeq. My proposed change is also not good, I realise, because `a * x` is not well-typed unless you abuse defeq. For me, the output of @[simps] is the correct lemma, I would advise using the correct constructors like `of_add` and `to_add` and the correct types like `multiplicative R` in the theorem *statements*. Feel free to abuse defeq all you like in the proofs.

Reply:
Just to be clear -- my current proposal with `mul_shift_spec` is to delete it and use `mul_shift_apply` instead.

Reply:
Oh wait -- I'm not even convinced that the lemma generated by `@[simps]` is even type correct in this strong sense. How about this:

```
@[simp] lemma mul_shift_apply {ψ : add_char R R'} {a : R} {x : multiplicative R} :
  (ψ.mul_shift a) x = ψ (of_add (a * to_add x)) := rfl

...

/-- If `n` is a natural number, then `mul_shift ψ n x = (ψ x) ^ n`. -/
lemma mul_shift_spec' (ψ : add_char R R') (n : ℕ) (x : multiplicative R) :
  mul_shift ψ n x = (ψ x) ^ n :=
by rw [mul_shift_apply, ← nsmul_eq_mul, of_add_nsmul, map_pow, of_add_to_add]

...

/-- The product of `mul_shift ψ a` and `mul_shift ψ b` is `mul_shift ψ (a + b)`. -/
lemma mul_shift_mul (ψ : add_char R R') (a b : R) :
  mul_shift ψ a * mul_shift ψ b = mul_shift ψ (a + b) :=
begin
  ext,
  simp only [monoid_hom.mul_apply, mul_shift_apply, add_mul, of_add_add, ψ.map_mul],
end
```
?
```

Reply:
Looks good.